Change neo GetAll.getNext to work with a delegate, not a buffer copy

gavin-norman-sociomantic opened the pull request:
This allows the storage engine to pass a slice of its contents to the
request handler's delegate, rather than requiring a copy into the handler's
buffer for each record. This is an optimisation that reduces CPU load
during iteration over large channels.

Part of https://github.com/sociomantic-tsunami/dhtnode/issues/19

nemanja-boric-sociomantic commented on src/dhtproto/node/neo/request/GetAll.d:
```diff
@@ -286,40 +282,46 @@ public abstract scope class GetAllProtocol_v0
             // Iterate over the channel and send each record to the client.
-            while ( this.outer.getNext(key, *this.outer.value_buffer) )
+            bool more = true;
+            while ( more )
```
`bool more; do { ... } while (more);`?

gavin-norman-sociomantic replied:
Updated.

nemanja-boric-sociomantic commented on src/dhtproto/node/neo/request/GetAll.d:
```diff
@@ -231,16 +227,16 @@ public abstract scope class GetAllProtocol_v0
 
-    abstract protected bool getNext ( out hash_t key, ref void[] value );
+    abstract protected bool getNext (
+        void delegate ( hash_t key, Const!(void)[] value ) dg );
```
I wonder if this delegate should return a boolean indicating if it want's to continue the operation, akin to `opApply`, but I can't think of the compelling reason right now, as unlike `opApply`, user is responsible of calling `getNext` for each next record.

nemanja-boric-sociomantic replied:
(in which case, returning false from the delegate would make getNext return false as well).

gavin-norman-sociomantic replied:
This delegate's function is just to fetch the current record; it has nothing to do with the decision about whether to continue or not.

nemanja-boric-sociomantic replied:
That's true right now, I'm just wondering if it makes sense to allow it to influence decision if the iteration should continue or not. Probably not, I wouldn't complicate it without any reason :+1:.

Commit:
Change neo GetAll.getNext to work with a delegate, not a buffer copy

This allows the storage engine to pass a slice of its contents to the
request handler's delegate, rather than requiring a copy into the handler's
buffer for each record. (The value retrieved from the storage engine is
immediately copied into the batch buffer, anyway, so the extra copy was
utterly pointless.)

This is an optimisation that reduces CPU load during iteration over large
channels.

Part of https://github.com/sociomantic-tsunami/dhtnode/issues/19

nemanja-boric-sociomantic commented on src/fakedht/neo/request/GetAll.d:
```diff
@@ -167,27 +167,29 @@ private scope class GetAllImpl_v0 : GetAllProtocol_v0
         auto ok = toHashT(str_key, key);
         assert(ok);
-        value = this.channel.get(key).dup;
+        auto value = this.channel.get(key);
```
Look, ma', no `dup`!
